compiler/src/typed/typecore.re

```diff
@@ -2690,29 +2756,96 @@ let report_error = (env, ppf) =>
+        "@[<v>@[<2>This expression has type@ %a@]@ %s@]",
+        type_expr,
+        typ,
+        "This is not a function; it cannot be applied.",
+      );
+    }
+  | Apply_too_many_arguments(typ, unused_tyargs) => {
+      reset_and_mark_loops(typ);
+      fprintf(ppf, "@[<v>@[<2>This function has type@ %a@]", type_expr, typ);
+      fprintf(ppf, "@ @[It is applied to too many arguments.@]@]");
+      let unused_optional_arguments =
+        List.filter_map(
+          ((l, _)) =>
+            switch (l) {
+            | Default({txt: name}) => Some(name)
+            | _ => None
+            },
+          unused_tyargs,
+        );
+      let oxford = (ppf, args) => {
+        let rec oxford = (ppf, args) =>
+          switch (args) {
+          | [] => ()
+          | [arg1, arg2] => fprintf(ppf, "%s, or %s", arg1, arg2)
+          | [arg1, ...args] =>
+            fprintf(ppf, "%s, ", arg1);
+            oxford(ppf, args);
+          };
+
+        switch (args) {
+        | [] => ()
+        | [arg] => fprintf(ppf, "%s", arg)
+        | [arg1, arg2] => fprintf(ppf, "%s or %s", arg1, arg2)
+        | _ => oxford(ppf, args)
+        };
+      };
+      switch (unused_optional_arguments) {
+      | [] => ()
+      | labels =>
         fprintf(
           ppf,
-          "@[<v>@[<2>This function has type@ %a@]",
-          type_expr,
-          typ,
-        );
+          "@ @[Did you mean to supply an optional argument with label %a?@]@]",
+          oxford,
+          unused_optional_arguments,
+        )
+      };
+    }
+  | Apply_too_few_arguments(args) => {
+      List.iter(((_, typ)) => reset_and_mark_loops(typ), args);
+      let print_arg = ((l, arg)) => {
+        reset_and_mark_loops(arg);
+        switch (l) {
+        | Unlabeled => fprintf(ppf, "%a", type_expr, arg)
+        | _ =>
+          fprintf(ppf, "%s: %a", qualified_label_name(l), type_expr, arg)
+        };
+      };
+      let rec print_args = (ppf, args) => {
+        switch (args) {
+        | [] => ()
+        | [arg] => print_arg(arg)
+        | [arg, ...rest] =>
+          print_arg(arg);
+          fprintf(ppf, ",@ ");
+          print_args(ppf, rest);
+        };
+      };
+      switch (args) {
+      | [arg] =>
         fprintf(
           ppf,
-          "@ @[It is applied to too many arguments;@ %s@]@]",
-          "maybe you forgot a `;'.",
-        );
+          "@[<hov>This function call is missing an argument of type %a@]",
+          print_args,
+          args,
+        )
       | _ =>
         fprintf(
           ppf,
-          "@[<v>@[<2>This expression has type@ %a@]@ %s@]",
-          type_expr,
-          typ,
-          "This is not a function; it cannot be applied.",
+          "@[<hov>This function call is missing arguments of type %a@]",
+          print_args,
+          args,
         )
       };
     }
+  | Apply_unknown_label(label, valid_labels) => {
+      fprintf(ppf, "This argument cannot be applied with label %s.", label);
```

Comment:
```suggestion
      fprintf(ppf, "This argument cannot be called with label %s.", label);
```

Reply:
The function is the thing being called, not the argument. Will reword:
```suggestion
      fprintf(ppf, "This argument cannot be supplied with label %s.", label);
```